redshift/redshift.go

```diff
@@ -71,149 +67,296 @@ func (c sortableColumns) Less(i, j int) bool { return c[i].Ordinal < c[j].Ordina
+	}
+	if err := rows.Err(); err != nil {
+		return nil, nil, fmt.Errorf("issue iterating over columns, err: %s", err)
+	}
+
+	// turn into Table struct
+	retTable := Table{
+		Name:    tableName,
+		Columns: cols,
+		Meta: Meta{
+			DataDateColumn: dataDateCol,
+			Schema:         schema,
+		},
+	}
+
+	// what's the last data in the table?
+	lastDataQuery := fmt.Sprintf("SELECT %s FROM %s.%s ORDER BY %s DESC LIMIT 1",
+		dataDateCol, schema, tableName, dataDateCol)
+	var lastData time.Time
+	if err = r.QueryRow(lastDataQuery).Scan(&lastData); err != nil {
+		return nil, nil, fmt.Errorf("issue running query: %s, err: %s", lastDataQuery, err)
+	}
+	return &retTable, &lastData, nil
+}
+
+func getColumnSQL(c ColInfo) string {
+	// note that we are relying on redshift to fail if we have created multiple sort keys
+	// currently we don't support that
+	defaultVal := ""
+	if c.DefaultVal != "" {
+		defaultVal = fmt.Sprintf("DEFAULT %s", c.DefaultVal)
+	}
+	notNull := ""
+	if c.NotNull {
+		notNull = "NOT NULL"
+	}
+	primaryKey := ""
+	if c.PrimaryKey {
+		primaryKey = "PRIMARY KEY"
+	}
+	sortKey := ""
+	if c.SortOrdinal == 1 {
+		sortKey = "SORTKEY"
+	}
+	distKey := ""
+	if c.DistKey {
+		distKey = "DISTKEY"
 	}
-	copySQL := `COPY "?"."?" FROM '?' WITH ? JSON '?' REGION '?' TIMEFORMAT 'auto' STATUPDATE ON COMPUPDATE ON %s`
-	copyStmt, err := r.Prepare(fmt.Sprintf(copySQL, credSQL))
+
+	return fmt.Sprintf(" %s %s %s %s %s %s %s", c.Name, typeMapping[c.Type], defaultVal, notNull, sortKey, primaryKey, distKey)
+}
+
+// RunCreateTable runs the full create table command in the provided transaction, given a
+// redshift representation of the table.
+func (r *Redshift) RunCreateTable(tx *sql.Tx, table Table) error {
+	var columnSQL []string
+	for _, c := range table.Columns {
+		columnSQL = append(columnSQL, getColumnSQL(c))
+	}
+	args := []interface{}{strings.Join(columnSQL, ",")}
+	// for some reason prepare here was unable to succeed, perhaps look at this later
+	createSQL := fmt.Sprintf(`CREATE TABLE "%s"."%s" (%s)`, table.Meta.Schema, table.Name, strings.Join(columnSQL, ","))
+	createStmt, err := tx.Prepare(createSQL)
 	if err != nil {
-		return err
+		return fmt.Errorf("issue preparing statement: %s", err)
+	}
+
+	log.Printf("Running command: %s with args: %v", createSQL, args)
+	_, err = createStmt.Exec()
+	return err
+}
+
+// RunUpdateTable figures out what columns we need to add to the target table based on the
+// input table, and completes this action in the transaction provided
+// Note: only supports adding columns currently, not updating existing columns or removing them
+func (r *Redshift) RunUpdateTable(tx *sql.Tx, targetTable, inputTable Table) error {
+	columnOps := []string{}
+	for _, inCol := range inputTable.Columns {
+		var existingCol ColInfo
+		// yes, n^2, but how many are we really going to have, and I want to keep the ordering
+		// ignore if there are extra columns
+		for _, targetCol := range targetTable.Columns {
```

Comment:
Alternative: construct a map of column_name -> *column. O(n) and maintains the order of inputTable.Columns. Might be unnecessary though.